File: Make Sum Divisible by P.py
# https://leetcode.com/problems/make-sum-divisible-by-p/?envType=daily-question&envId=2024-10-03
# Make Sum Divisible by P

class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:

        total_sum = sum(nums)
        target = total_sum % p

        if target == 0:
            return 0

        n = len(nums)
        prefix_sum = 0
        min_len = n
        prefix_mod_map = {0: -1}

        for i in range(n):
            prefix_sum += nums[i]
            mod = prefix_sum % p
            
            needed_mod = (mod - target) % p
            if needed_mod in prefix_mod_map:
                min_len = min(min_len, i - prefix_mod_map[needed_mod])
            
            prefix_mod_map[mod] = i

        return min_len if min_len < n else -1

File: Make Sum Divisible by P.js
// https://leetcode.com/problems/make-sum-divisible-by-p/submissions/1843530239/?envType=daily-question&envId=2025-11-30
// Make Sum Divisible by P

/**
 * @param {number[]} nums
 * @param {number} p
 * @return {number}
 */
var minSubarray = function(nums, p) {
    
    let totalSum = 0;
    
    for (const num of nums) {
        totalSum = (totalSum + num) % p;
    }
    
    const targetRemainder = totalSum;
    
    if (targetRemainder === 0) return 0;

    const modMap = new Map();
    modMap.set(0, -1);
    
    let currentSum = 0;
    let minLen = nums.length;
    
    for (let i = 0; i < nums.length; i++) {
        currentSum = (currentSum + nums[i]) % p;
        
        const needed = (currentSum - targetRemainder + p) % p;
        
        if (modMap.has(needed)) {
            const prevIndex = modMap.get(needed);
            minLen = Math.min(minLen, i - prevIndex);
        }
        
        modMap.set(currentSum, i);
    }
    
    return minLen < nums.length ? minLen : -1;
};